### ISSUE TITLE
How to render gaussian trained from official implemention.

### ISSUE THREAD

**Issue by jiang25262** [hbb1/torch-splatting#3]
I change the renderer size from 256*256 to H*W and add focal to camera. But the output is all 1.0 in render_pkg["render"]

**Comment by hbb1** [comment]
Yes, the resolution is hardcoded and should be rectified. Giving correct resolution and camera the render should work. I'm not sure where can cause this error, can you give more descriptions?

**Comment by xiazhi1** [comment]
> I change the renderer size from 256_256 to H_W and add focal to camera. But the output is all 1.0 in render_pkg["render"]
Awesome work !
Now , I am doing the same work like @jiang25262  , but now I don't really know how to add focal to camera  and why we should change the renderer size , can you give me some suggestions? Looking forward for your apply !
@jiang25262 @hbb1

**Comment by hbb1** [comment]
@xiazhi1  you should  move add one  line of code here      https://github.com/hbb1/torch-splatting/blob/main/gaussian_splatting/gauss_render.py#L176 

self.pix_coord = torch.stack(torch.meshgrid(torch.arange(camera.image_width), torch.arange(camera.image_height), indexing='xy'), dim=-1).to('cuda')

**Comment by xiazhi1** [comment]
> @xiazhi1 you should add one line of code here https://github.com/hbb1/torch-splatting/blob/main/gaussian_splatting/gauss_render.py#L176
> 
> self.pix_coord = torch.stack(torch.meshgrid(torch.arange(camera.image_width), torch.arange(camera.image_height), indexing='xy'), dim=-1).to('cuda')

@hbb1  Thanks for your reply, I have tried it but it still fall in https://github.com/hbb1/torch-splatting/blob/e2d78419edca7847cc937376ed893e939f76a572/gaussian_splatting/gauss_render.py#L213C11-L213C11 . I believe the error is caused by the size mismatch between reshaped tile_color and subarea of self.render_color. Now I am not sure why the size mismatch , so I want to try to use the 360v2 dataset(Gaussian_Splatting offical dataset) which is a colmap dataset to run torch-splatting to test the tile size? Do you think this is  useful ? And How can I change the dataset to 360v2 dataset ?   Looking forward for your apply ! Also, Wish you a Happy New Year！

**Comment by xiazhi1** [comment]
> @xiazhi1 you should add one line of code here https://github.com/hbb1/torch-splatting/blob/main/gaussian_splatting/gauss_render.py#L176
> 
> self.pix_coord = torch.stack(torch.meshgrid(torch.arange(camera.image_width), torch.arange(camera.image_height), indexing='xy'), dim=-1).to('cuda')

@hbb1 Happy New Year! Thanks for your timely reply, now I think I already finish change the render size, but I failed with  CUDA OOM in the render part even though it just an interation when I am debugging, I want to know approximately how much cuda memory does your model occupy? I use the 360v2 dataset whose image size is 1063*1600?

**Comment by xiazhi1** [comment]
> @xiazhi1 you should add one line of code here https://github.com/hbb1/torch-splatting/blob/main/gaussian_splatting/gauss_render.py#L176
> 
> self.pix_coord = torch.stack(torch.meshgrid(torch.arange(camera.image_width), torch.arange(camera.image_height), indexing='xy'), dim=-1).to('cuda')

@hbb1 Happy New Year ! Thanks a lot for your timely reply, I have tried it and successfully change the renderer size to fit the official implemention's 360v2 dataset whose image size is 1063*1600, but now I failed because of CUDA OOM， I wonder whether it is caused by the bigger image size or more python loop? And I tried to increse tilesize from 16 to 512 to decrease the number of python loop, but it still have same problem of CUDA OOM , Where do you think the problem might be? If I want to try to use the dataset of yours to run on official code, What do I need to pay attention to？ Looking forward for your reply!

**Comment by xiazhi1** [comment]
> Hi, @xiazhi1 Please refer to the code snippet at this link: https://github.com/hbb1/torch-splatting/blob/main/gaussian_splatting/gauss_render.py#L200C77-L200C77.
> 
> As we aim to calculate the blending weight derived from the Gaussian, it's needed to generate a matrix of dimensions (T, P, 2), where T represents the tile size, and P means the number of points. When dealing with substantial tile sizes or a high volume of points, there's a possibility of encountering Out Of Memory (OOM) errors.

ok, thanks a lot ! I will check it.

**Comment by wb014** [comment]
Thanks for the torch implementation. However, I still fail to render correct result with your implementation and official trained point clound. The rendered images somehow carry noise and hole. Do you have any ideas? Looking forward to your reply. Rendered images are attached below(left one is rendered with official code). 
![Uploading rendered_imgs.png…]()

**Comment by hbb1** [comment]
@wb014  Hi, I think another important aspect may be the learning rate. Note that the original implementation assigns different attributes with different learning rate. Indeed, their github readme mentioned that the learning rate is of great important. Aligning the learning rate with the original implementation may help.

**Comment by wb014** [comment]
> @wb014 Hi, I think another important aspect may be the learning rate. Note that the original implementation assigns different attributes with different learning rate. Indeed, their github readme mentioned that the learning rate is of great important. Aligning the learning rate with the original implementation may help.

sorry, I mean I use this torch-splatting to render the 3d gaussians trained from official 3dgs code. With the same gaussians, the left image is rendered with official code, and the right image is rendered with your implementation. I can't locate the mistake

**Comment by hbb1** [comment]
@wb014  I see. But I have no idea right now. Will you be able to upload your ply file and related code so I can have a check later?

**Comment by hbb1** [comment]
@wb014  Thank you. I will let you know if I have any progress.

**Comment by wb014** [comment]
@hbb1 Have you received the email? I've sent you the code and data. And I found two mistakes or difference compared with the official code. 
First one is in function build_color(), there is a normalize in the official code. torch-splatting can be quick fixed with `rays_d = rays_d / rays_d.norm(dim=-1, keepdim=True)` following `rays_d = means3D - rays_o` .
Second one is a bug in calculation of transmittance. `T = torch.cat([torch.ones_like(alpha[:,:1]), 1-alpha[:,1:]], dim=1).cumprod(dim=1)` may should be fixed to `T = torch.cat([torch.ones_like(alpha[:, :1]), 1 - alpha[:, :-1]], dim=1).cumprod(dim=1)`
However, after fixing the above, there is still something wrong when rendering 3d gaussians trained from official code with torch-splatting. The upper part of the rendered image is in blur. As can be seen below.
<img width="1415" alt="render_0301" src="https://github.com/hbb1/torch-splatting/assets/53701627/718f328e-c6d1-41b7-a6d3-cbb8a7ba0a4b">

**Comment by hbb1** [comment]
@wb014  Hi, thank you for point this. The bottom region looks really good now so now the problem comes to the upper part. 
I  further confirm the correctness without using tile. 

<img width="518" alt="image" src="https://github.com/hbb1/torch-splatting/assets/57170935/ed41e8c6-20c8-4da7-8ac9-2716364ce919">

Here is the code:

    def gaussian_splatting(means3D, scales, quats, colors, intrins, viewmat, projmat, opacity):
        projmat = torch.zeros(4,4).cuda()
        projmat[:3,:3] = intrins
        projmat[-1,-2] = 1.0
        projmat = projmat.T

        mean_ndc, mean_view, in_mask = projection_ndc(means3D, viewmatrix=viewmat, projmatrix=projmat)

        depths = mean_view[:,2]
        mean_coord_x = mean_ndc[..., 0]
        mean_coord_y = mean_ndc[..., 1]
        # mean_coord_x = ((mean_ndc[..., 0] + 1) * camera.image_width - 1.0) * 0.5
        # mean_coord_y = ((mean_ndc[..., 1] + 1) * camera.image_height - 1.0) * 0.5

        means2D = torch.stack([mean_coord_x, mean_coord_y], dim=-1)
        scales = torch.cat([scales[..., :2], scales[..., -1:]*1e-2], dim=-1)
        cov3d = build_covariance_3d(scales, quats)

        W, H = intrins[0,-1] * 2, intrins[1,-1] * 2
        fx, fy = intrins[0,0], intrins[1,1]
        tan_fovx = W / (2 * fx)
        tan_fovy = H / (2 * fy)
        cov2d = build_covariance_2d(means3D, cov3d, viewmat, tan_fovx, tan_fovy, fx, fy)
        radii = get_radius(cov2d)
        
        # Rasterization
        pix = torch.stack(torch.meshgrid(torch.arange(H), torch.arange(W), indexing='xy'), dim=-1).to('cuda').flatten(0,-2)
        sorted_conic = cov2d.inverse() # inverse of variance
        dx = (pix[:,None,:] - means2D[None,:]) # B P 2
        dist2 = dx[:, :, 0]**2 * sorted_conic[:, 0, 0] + dx[:, :, 1]**2 * sorted_conic[:, 1, 1]+ dx[:,:,0]*dx[:,:,1] * sorted_conic[:, 0, 1]+ dx[:,:,0]*dx[:,:,1] * sorted_conic[:, 1, 0]
        depth_acc = depths[None].expand_as(dist2)

        dist2 = dist2.T
        dist2 = 100 * (dist2 - 1).clamp(min=0)
        colors = colors.reshape(-1,1,colors.shape[-1])
        depth_acc = depth_acc.T[..., None]
        depth_acc = depth_acc.repeat(1,1,1)
        alpha = torch.exp(-0.5*dist2)[..., None] * opacity
        T = torch.cat([torch.ones_like(alpha[-1:]), (1-alpha).cumprod(dim=0)[:-1]], dim=0)
        image = (T * alpha * colors).sum(dim=0).reshape(H.long().item(), W.long().item(), colors.shape[-1])
        alphamap = (T * alpha).sum(dim=0).reshape(H.long().item(), W.long().item(), 1)
        image = image + (1-alphamap)
        depthmap = (T * alpha * depth_acc).sum(dim=0).reshape(H.long().item(), W.long().item(), 1)
        dist2 = dist2.min(0)[0].reshape(H.long(),W.long()).sqrt()
        return image, depthmap, center, radii

So the problem may lie in the tile? (no sure)
btw, your email was well received but having a tight schedule, I may not be able to check it recent days :)

**Comment by wb014** [comment]
@hbb1 I tried another dataset. And I found something weird and interesting.
<img width="1427" alt="fence_error" src="https://github.com/hbb1/torch-splatting/assets/53701627/10ea88bc-e6f0-4546-beae-aafef1d83bf5">
The fence cannot be observed from this view, but it is rendered in wrond position and direction. Meanwhile, a tire of the truck is rendered correctly.
It's hard to locate the problem, maybe not the tile issue, but in 3dg projection to 2d space. For the reason that the gaussians constructing fence should not be seen from this view. 

I figure it out! 3dg should be pre-filtered according to the depth, although I remember that 'prefilter' is set to False in official implementation. Anyway it renders correctly now.

**Comment by hbb1** [comment]
Yes, I tested the implementation only for bounded scene before. We should indeed filter those splats behind the camera to support such cases. I will fix it soon.

**Comment by weizizhuang** [comment]
Nice work! I'm really looking forward to a pure pytorch version

**Comment by hbb1** [comment]
Fixed  here https://github.com/hbb1/torch-splatting/commit/85c362152883cda066d6a1cc949c2bab7e2cc1c7

**Comment by hi-zhangjie** [comment]
> @hbb1 I tried another dataset. And I found something weird and interesting. <img alt="fence_error" width="1427" src="https://private-user-images.githubusercontent.com/53701627/309439176-10ea88bc-e6f0-4546-beae-aafef1d83bf5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjA3NTIwODAsIm5iZiI6MTcyMDc1MTc4MCwicGF0aCI6Ii81MzcwMTYyNy8zMDk0MzkxNzYtMTBlYTg4YmMtZTZmMC00NTQ2LWJlYWUtYWFmZWYxZDgzYmY1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzEyVDAyMzYyMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWM4YzM1MmEzZWZiZjgwMjIwYjhmZGRiMjVlYWI3Yzg3NzgxMWY1YTA1NGI4OTlmN2I3MThkN2VmM2FiNjAwYjMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.46SrNwFf9CcElieS2XpLbKJDlzkFJXFui4MZAlDi5wc"> The fence cannot be observed from this view, but it is rendered in wrond position and direction. Meanwhile, a tire of the truck is rendered correctly. It's hard to locate the problem, maybe not the tile issue, but in 3dg projection to 2d space. For the reason that the gaussians constructing fence should not be seen from this view.
> 
> I figure it out! 3dg should be pre-filtered according to the depth, although I remember that 'prefilter' is set to False in official implementation. Anyway it renders correctly now.

@wb014 Hi，Do you have any pre-filtering based on depth? Currently, it renders correctly？ Can you provide the rendering script, thx.
